Clean up model_storm_module and bound azimuthal wind at 0

bolliger32 opened the pull request:
The main change from this PR is to close #465 by clipping azimuthal wind speed at 0 (no negative values). In addition, I cleaned up the model_storm_module file a bit by pulling out common pre- and post-processing functionality into separate subroutines (including this bounding at 0 function). I noticed a few errors in some of the wind models that I noted in #466 but didn't address as my knowledge of these wind models is limited

mandli commented on src/2d/shallow/surge/model_storm_module.f90:
```diff
@@ -383,6 +383,92 @@ subroutine get_storm_data(t, storm, location, velocity, max_wind_radius,  &
+
+    ! 2. Convert wind speed at 10 m to top of atmospheric boundary layer
+    ! ==========================================================================
+    subroutine adjust_max_wind(tv, mws, mod_mws)
```
Maybe make this a `pure function` so that it can be inlined automatically?

bolliger32 replied:
OK i added the pure modifier. Was that what you meant? Or is there something deeper. I wasn't able to figure out much about pure procedures online so I'm not really sure what I'm doing. I dont' think this can easily be made into a function? B/c it's modifying tv in addition to producing mod_mws

mandli replied:
Did it actually compile then?  I thought `pure` could only be added to a function.

bolliger32 replied:
it did...but i'm wondering if it just ignored the pure modifier? Here's what I was using to try to figure out how pure works: http://www.lahey.com/docs/lfpro78help/F95ARPURE.htm. Suggests that it can be applied to subroutines, but seems like these ones don't actually meet the criteria

mandli commented on src/2d/shallow/surge/model_storm_module.f90:
```diff
@@ -383,6 +383,92 @@ subroutine get_storm_data(t, storm, location, velocity, max_wind_radius,  &
+
+    ! 5. Apply distance ramp to limit scope
+    ! ==========================================================================
+    subroutine post_process_wind_estimate(maux, mbc, mx, my, i, j, wind, aux, &
```
Again, is it possible to do with a `pure` function?

bolliger32 replied:
Same thing here. I just slapped pure on there, since changing the subroutine to a function would be tricker since it modifies a couple of values.

mjberger commented:
coming in late to this discussions. But I"ve had pure functions not work in openmp with multiple threads. Do you understand this?

In my case it was an "iadd" pure function, and I had to put it back to being a statement function.

 — Marsha

> On Aug 11, 2020, at 4:34 PM, Kyle Mandli <notifications@github.com> wrote:
> 
> 
> @mandli commented on this pull request.
> 
> Just one comment regarding the resulting new functions. Sometimes pure functions are not actually faster but at least a compiler can use it if it has the capability.
> 
> In src/2d/shallow/surge/model_storm_module.f90 <https://github.com/clawpack/geoclaw/pull/467#discussion_r468847712>:
> 
> > @@ -383,6 +383,92 @@ subroutine get_storm_data(t, storm, location, velocity, max_wind_radius,  &
>      end subroutine get_storm_data
>  
>  
> +    ! ==========================================================================
> +    ! A set of common functions applied before calculating the circular component of 
> +    ! wind speed using a particular wind model.
> + 
> +    ! 1. Adjust max wind speed by removing translational speed (bounding at 0).
> +
> +    ! 2. Convert wind speed at 10 m to top of atmospheric boundary layer
> +    ! ==========================================================================
> +    subroutine adjust_max_wind(tv, mws, mod_mws)
> Maybe make this a pure function so that it can be inlined automatically?
> 
> In src/2d/shallow/surge/model_storm_module.f90 <https://github.com/clawpack/geoclaw/pull/467#discussion_r468848343>:
> 
> > +    ! has been calculated given a particular wind model:
> + 
> +    ! 1. Determine translation speed that should be added to final storm wind speed.
> +    !    This is tapered to zero as the storm wind tapers to zero toward the eye of the 
> +    !    storm and at long distances from the storm. Avoid divide-by-0 error
> +
> +    ! 2. Convert wind velocity from top of atmospheric boundary layer to wind
> +    !    velocity at 10 m above the earth's surface.
> +
> +    ! 3. Convert from 1 minute averaged winds to 10 minute averaged winds
> +
> +    ! 4. Add back in translational velocity
> +
> +    ! 5. Apply distance ramp to limit scope
> +    ! ==========================================================================
> +    subroutine post_process_wind_estimate(maux, mbc, mx, my, i, j, wind, aux, &
> Again, is it possible to do with a pure function?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/clawpack/geoclaw/pull/467#pullrequestreview-465407122>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAGUGC7AE2FILRMEIGUDCSLSAGTNLANCNFSM4P3NPXMQ>.
>

mandli commented:
Nope, not sure why that would happen.  Do you know what the situation was?

mjberger commented:
Nope. But it completely surprised me. I had to track it down with print statements.

 — Marsha

> On Aug 11, 2020, at 4:38 PM, Kyle Mandli <notifications@github.com> wrote:
> 
> 
> Nope, not sure why that would happen. Do you know what the situation was?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/clawpack/geoclaw/pull/467#issuecomment-672266745>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAGUGC5WCX2Q76KCUX5V6ILSAGT2RANCNFSM4P3NPXMQ>.
>

mandli commented:
Ug, Fortran

bolliger32 commented:
@mandli @mjberger I'll give pure functions a try... with the caveat that I don't really know what I'm doing in fortran :) 

I wonder if it's possible to make the ike test use openmp. Not that it would catch every potential gotcha case, but might be one line of defense.

bolliger32 commented:
And just to confirm i did run the ike test on my machine with openMP and it didn't seem to cause any issues.

bolliger32 commented:
OK I went down a bit of a rabbit hole into these vortex models in trying to clean this up. I've updated a few more things:

General cleaning
-----------------
- I pulled out the Holland B parameter calculation into its own pure function
- I pulled out the polar coordinate calculation into its own pure function
- I pulled out the pressure difference calculation into its own pure function (which for some models was bounding at 100Pa but not for all models)
- I added some references for the models in the docstrings

Model updating
----------------
- I updated several models where the full max observed wind was being used rather than the max of the circular wind speed
- I uncommented a line for calculating "x" in the Holland 2010 model, so it should now work
- I did NOT touch the CLE model as this is above my pay grade :) 
- I corrected Holland 2010 to use surface level obs (while all the others used the gradient level winds)
- I updated the SLOSH model to use the formula for combining translational speed and circular speed that is actually employed in SLOSH
- I provided an updated estimation of alpha in the modified rankine model based on a 2005 paper
- I finished up what looked like a partially finished Willoughby model

These are a lot of changes, so I'm not offended if you want me to chuck some of them. It's probably worth @hudaquresh taking a look, since I'm sure you're pretty familiar with whats going on here and can let me know if/where i screwed up :) 

These changes should additionally close #466

mandli commented:
Sorry, been super busy with the beginning of a semester of some sort.  Given that @hudaquresh is studying for an exam these days so I took a look.  All the changes look good to me but it's hard to give a general commentary given we do not have full test coverage on all of these.

In terms of the CLE model though, my guess is that it does not need the ramping function as it actually limits to zero unlike most of the others.  We should open an issue if there isn't one already to figure out the CLE model cleanup.
